gapis/resolve/memory.go

```diff
@@ -73,7 +73,33 @@ func Memory(ctx context.Context, p *path.Memory) (*service.Memory, error) {
 			interval.Merge(&writes, rng.Window(r).Span(), false)
 		}
 	}
-	api.MutateCmds(ctx, s, nil, cmds[len(cmds)-1])
+	// api.MutateCmds(ctx, s, nil, cmds[len(cmds)-1])
+
+	endCmd := cmds[len(cmds)-1]
+	if syncedApi, ok := endCmd.API().(sync.SynchronizedAPI); ok {
+		requestSubCmdIdx := api.SubCmdIdx(p.After.Indices)
+		syncedApi.MutateSubcommands(ctx, api.CmdID(cmdIdx), endCmd, s,
+			func(s *api.State, subCommandIndex api.SubCmdIdx, cmd api.Cmd) {
+				// Clear the read/write list before grabbing the observations for the
+				// requested subcommand
+				if requestSubCmdIdx.Equals(subCommandIndex) {
+					reads = memory.RangeList{}
```

Comment:
Instead of doing the work to add the observation, then remove it repeatedly, bind the `OnRead` and `OnWrite` handlers once you've reached the requested subcommand?

You can remove duplication by wrapping the handlers in a function called from the post, and the non-syncedApi branch:

```go
	listenToReadWrite := func() {
		pool.OnRead = func(rng memory.Range) {
			if rng.Overlaps(r) {
				interval.Merge(&reads, rng.Window(r).Span(), false)
			}
		}
		pool.OnWrite = func(rng memory.Range) {
			if rng.Overlaps(r) {
				interval.Merge(&writes, rng.Window(r).Span(), false)
			}
		}
	}
```